modelence, auth-ui: Make email verification mandatory

omegascorp opened the pull request:
* Send email verification on sign-up
* Send email verification on sign-in if the email is not verified, and don't allow them to sign in
* If there is no email provider, it allows signing in without confirmation. I think this will work well for the initial setup when the email is not set up yet

<img width="1512" height="757" alt="Screenshot 2025-08-06 at 11 07 52" src="https://github.com/user-attachments/assets/d400e881-144e-46eb-a8fa-ed52cf24621f" />

<img width="1512" height="757" alt="Screenshot 2025-08-06 at 11 16 33" src="https://github.com/user-attachments/assets/5e089dda-8209-4f13-860f-46d4904cde80" />

artahian commented on packages/auth-ui/src/components/SignupForm.tsx:
```diff
@@ -124,6 +130,27 @@ export function SignupForm({
+            We've sent you a verification email. Please check your inbox and click the verification link to activate your account.
+          </p>
+          <p className="text-sm text-gray-500 dark:text-gray-500">
+            Don't see the email? Check your spam folder.
```
@omegascorp I guess that part is ok, but what about the resend part?

Right now, if you don't click on the email before it expires or somehow don't receive the email (or accidentally delete it), you seem to be forever locked as a user, and can't log in, but at the same time you also can't sign up with that same email anymore.

I think at the minimum we need to allow users to request the verification email again. Perhaps we should add a separate verification page (or this can be a different state of the login form component) that shows the same card, saying something like:

"You need to verify your email before logging in. We've sent an email to xxx@yyy.zz, click the link, etc, etc. 

[Resend Button]"

this component can be used for both cases, at any time. After signup, we can redirect to the verification page (or to the login form, but immediately to this state that shows the verification card). And at any time, if the user later comes back in and tries to log in, they will see the same card, with the same Resend button option.

This way we have a single consistent card, regardless of whether it's right after signup or coming back later, and there will always be a way to resend the email without being completely locked out.

omegascorp replied:
> Send email verification on sign-in if the email is not verified, and don't allow them to sign in

Aram Jan, please check the second screenshot. It sends an email verification automatically when you are trying to sign in with an email that is not verified.

omegascorp commented:
@artahian, I've added a rate limit on login and a separate one on verification. I've also moved sign-up verification email up, since without that you can, for example, brute-force which emails we consider disposable, or check which emails are registered.

artahian commented on packages/modelence/src/auth/user.ts:
```diff
@@ -59,6 +59,26 @@ export default new Module('_system.user', {
+    window: time.days(1),
+    limit: 200,
+  }, {
+    bucket: 'verification',
```
Shouldn't the bucket be the user for the verification, since we have the user at that point already?

artahian commented on packages/modelence/src/auth/login.ts:
```diff
@@ -19,13 +31,39 @@ export async function handleLoginWithPassword(args: Args, { user, session }: Con
+    throw new Error("Your email address hasn't been verified yet. We've sent a new verification email to your inbox.");
+  }
+
+  const passwordHash = userDoc.authMethods?.password?.hash;
```
Shouldn't this check be before the email verified checks? If the user is providing a wrong password, we should just reject the login, right? Otherwise you can still brute force unverified emails and trigger the verification emails even though you don't even have the password.

artahian commented on packages/modelence/src/auth/signup.ts:
```diff
@@ -71,38 +59,19 @@ export async function handleSignupWithPassword(args: Args, { user, connectionInf
-      subject: getEmailConfig()?.verification?.subject || 'Verify your email address',
-      text: textContent,
-      html: htmlTemplate,
+  if (ip) {
+    await consumeRateLimit({
+      bucket: 'signup',
+      type: 'ip',
+      value: ip,
     });
```
I think this should still be in its original position, before inserting the user document? Otherwise we still do the signup, create the user, and only then check the rate limit, which is too late.

omegascorp commented:
@artahian comments are addressed
